app/src/main/java/org/mozilla/fenix/components/metrics/AppStartupTelemetry.kt

```diff
@@ -0,0 +1,147 @@
+        setOnCreateData(safeIntent, hasSavedInstanceState, true)
+    }
+
+    fun onHomeActivityOnRestart() {
+        // we are not setting [Source] in this method since source is derived from an intent.
+        // therefore source gets set in onNewIntent().
+        onRestartData = Pair(HOT, false)
+    }
+
+    fun onHomeActivityOnNewIntent(safeIntent: SafeIntent) {
+        // we are only setting [Source] in this method since source is derived from an intent].
+        // other metric fields are set in onRestart()
+        onNewIntentData = getStartupSourceFromIntent(safeIntent, false)
+    }
+
+    private fun setOnCreateData(
+        safeIntent: SafeIntent,
+        hasSavedInstanceState: Boolean,
+        isExternalAppBrowserActivity: Boolean
+    ) {
+        onCreateData = AppAllStartup(
+            getStartupSourceFromIntent(safeIntent, isExternalAppBrowserActivity),
+            getAppStartupType(),
+            hasSavedInstanceState
+        )
+        wasAppCreateCalledBeforeActivityCreate = false
+    }
+
+    private fun getAppStartupType(): Type {
+        return if (wasAppCreateCalledBeforeActivityCreate) COLD else WARM
+    }
+
+    private fun getStartupSourceFromIntent(
+        intent: SafeIntent,
+        isExternalAppBrowserActivity: Boolean
+    ): Source {
+        return when {
+            isExternalAppBrowserActivity -> CUSTOM_TAB
+            intent.isLauncherIntent -> APP_ICON
+            intent.action == Intent.ACTION_VIEW -> LINK
+            // one of the unknown case is app switcher, where we go to the recent tasks to launch
+            // Fenix.
+            else -> UNKNOWN
+        }
+    }
+
+    /**
+     * The reason we record metric on resume is because we need to wait for onNewIntent(), and
+     * we are not guaranteed that onNewIntent() will be called before or after onStart() / onRestart().
+     * However we are guaranteed onResume() will be called after onNewIntent() and  onStart(). Source:
+     * https://developer.android.com/reference/android/app/Activity#onNewIntent(android.content.Intent)
+     */
+    fun onHomeActivityOnResume() {
+        recordMetric()
+    }
+
+    private fun recordMetric() {
+        if (!isMetricRecordedSinceAppWasForegrounded) {
+            val appAllStartup: AppAllStartup = if (onCreateData != null) {
+                onCreateData!!
+            } else {
+                mergeOnRestartAndOnNewIntentIntoStartup()
+            }
+            metrics.track(appAllStartup)
+            isMetricRecordedSinceAppWasForegrounded = true
+        }
+        // we don't want any weird previous states to persist on our next metric record.
+        onCreateData = null
+        onNewIntentData = null
+        onRestartData = null
+    }
+
+    private fun mergeOnRestartAndOnNewIntentIntoStartup(): AppAllStartup {
+        return if (onRestartData != null && onNewIntentData != null) {
```

Comment:
I think you can replace this if/else block with elvis operators:
```kotlin
return AppAllStartup(onNewIntentData ?: UNKNOWN, onRestartData.first ?: ERROR, onRestartData.second ?: false)
```

That being said, you'll need to adjust for having `hasSavedInstance` being an optional value if you follow my suggestion above.

Reply:
ah keep forgetting about the elvis operator! thanks!